### PR TITLE
Fixes for macos

### PR DESCRIPTION
**Pull request by hamishmack**
###### Motivation for this change

Gets WebKitGTK+, GtkSourceView and gtk-mac-integration working on MacOS

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [X] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/libraries/webkitgtk/2.14.nix [hunk 1]
```diff
@@ -18,12 +18,36 @@ stdenv.mkDerivation rec {
...
 
-  preConfigure = "patchShebangs Tools";
+  preConfigure = ''
+    patchShebangs Tools
+  '' + optionalString (stdenv.isDarwin) ''
+    patch -p0 < ${./PR-152650-2.patch}
+    patch -p0 < ${./PR-153138.patch}
+    patch -p0 < ${./PR-157554.patch}
+    patch -p0 < ${./PR-157574.patch}
+  '';
```
Comment (periklis) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: Put these in `patches` and do not use the patch util manually.

File: pkgs/development/libraries/webkitgtk/2.14.nix [hunk 2]
```diff
@@ -18,12 +18,36 @@ stdenv.mkDerivation rec {
...
+        Source/JavaScriptCore/CMakeFiles/LLIntOffsetsExtractor.dir/link.txt \
+        Source/JavaScriptCore/shell/CMakeFiles/jsc.dir/link.txt \
+        Source/JavaScriptCore/shell/CMakeFiles/testb3.dir/link.txt \
+        Source/WebKit2/CMakeFiles/DatabaseProcess.dir/link.txt \
+        Source/WebKit2/CMakeFiles/NetworkProcess.dir/link.txt \
+        Source/WebKit2/CMakeFiles/webkit2gtkinjectedbundle.dir/link.txt \
+        Source/WebKit2/CMakeFiles/WebProcess.dir/link.txt
+    sed -i "s|\.\./\.\./lib/libWTFGTK\.a|-Wl,-all_load ../../lib/libWTFGTK.a|" \
+        Source/JavaScriptCore/CMakeFiles/JavaScriptCore.dir/link.txt
+  '';
```
Comment (periklis) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: If possible try to use of `substituteInPlace`

File: pkgs/development/libraries/webkitgtk/2.14.nix [hunk 3]
```diff
@@ -37,24 +61,40 @@ stdenv.mkDerivation rec {
   cmakeFlags = [
   "-DPORT=GTK"
   "-DUSE_LIBHYPHEN=0"
-  "-DENABLE_GLES2=ON"
-  ];
+  ] ++ (if stdenv.isDarwin then [
```
Comment (periklis) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: Use of `stdenv.lib.optionals (stdenv.isDarwin) [...]` is a much nicer syntax that avoids parenthesis nesting and plays well with additions for other platforms.

File: pkgs/development/libraries/webkitgtk/2.14.nix [hunk 4]
```diff
@@ -37,24 +61,40 @@ stdenv.mkDerivation rec {
...
-  buildInputs = [
-    gtk2 wayland libwebp enchant libnotify gnutls pcre nettle libidn
+  buildInputs = libintlOrEmpty ++ [
+    gtk2 libwebp enchant libnotify gnutls pcre nettle libidn
     libxml2 libsecret libxslt harfbuzz libpthreadstubs libtasn1 p11_kit
     gst-plugins-base libxkbcommon epoxy at_spi2_core
   ] ++ optional enableGeoLocation geoclue2
-    ++ (with xlibs; [ libXdmcp libXt libXtst ]);
+    ++ (with xlibs; [ libXdmcp libXt libXtst ])
+    ++ (if stdenv.isDarwin then [ libedit readline mesa ] else [ wayland ]);
```
Comment (periklis) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: As above use `stdenv.lib.optionals` and make the platform explicit for `wayland`.

**Review by periklis — COMMENTED** [review]
Thx for the great work, i've proposed some enhnacements.

File: pkgs/development/libraries/webkitgtk/2.14.nix (hunk 3 above)
Comment (Mic92) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: might be an exception here, because of the `else` branch.

**Comment by hamishmack** [comment]
Ok I think I have the enhancements right now. @Mic92 I used `isLinux` for the `else` case.

File: pkgs/development/libraries/webkitgtk/2.14.nix [hunk 5]
```diff
@@ -32,29 +47,54 @@ stdenv.mkDerivation rec {
 
   # see if we can clean this up....
 
-  patches = [ ./finding-harfbuzz-icu.patch ];
+  patches = [ ./finding-harfbuzz-icu.patch ]
+  	++ optionals stdenv.isDarwin [
+    ./PR-152650-2.patch
```
Comment (LnL7) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: Can you use `fetchpatch` for these?

Comment (hamishmack) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: I don't think we can fetch them directly from macports if that is what you were suggesting.  I had to modify them slightly.  The macports patches are designed to work with `patch -p0`.  I had to add an extra directory to the paths to make them work with `patches` in nix (https://github.com/NixOS/nixpkgs/pull/24700/commits/90f4361fa0dc783c91b8bf2cdf1aefb204ce4893#diff-7e2407d988a384c067352687b99c845e is the commit with that change).

**Comment by matthewbauer** [comment]
So if this gets merged I think we can probably close #14284 because it looks like it's roughly the same purpose.

**Comment by matthewbauer** [comment]
Also: make sure your commit messages meet [the contributing guideles](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

File: pkgs/development/libraries/webkitgtk/2.14.nix (hunk 5 above)
Comment (LnL7) [inline pkgs/development/libraries/webkitgtk/2.14.nix]: Allrigth

**Comment by hamishmack** [comment]
@matthewbauer I have merged the `Use nix idiomatic patch and replace functions` commit into the first commit and added some detail of on the second commit message.

**Comment by LnL7** [comment]
It's preferred to format the first like `webkgitgtk: fix 2.14 on macOS`. Not mandatory, but it makes it easier for people to read the changelog if everything is formatted consistently.

**Comment by hamishmack** [comment]
I have moved the package names to the start of the commit messages.  Split the request into three commits (one for each package).  I also moved the `gtk` from `buildInputs` to `propergatedBuildInputs` in the `gtk-mac-integration` package.

**Review by LnL7 — APPROVED** [review]

**Pull request merged**